grammar.js

```diff
@@ -959,14 +967,10 @@ module.exports = grammar({
       ),
     ),
 
-    // TODO arbitrary dollar quoting, mostly ensuring that the initial and terminal quote
-    // delimiters match, will require an external scanner
-    // https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners
-    dollar_quote: () => choice(
-      '$$',
-      '$function$',
-      '$body$'
-    ),
+    // This is only used in create function statement, it is not needed to check
+    // the start tag match the end one. The usage of this syntax in other
+    // context is done by _dollar_string.
```

Comment:
I don't understand the only thing I said is `$._dollar_string` and `$.dollar_quote` don't interfere and there is no need to check the closing `$.dollar_quote` to match.

And for format I don't understand, here is how it pase:
```sql
select format($$hello %s$$, format($$all the %s$$, $$world$$));
```

```
(program
 (statement
  (select (keyword_select)
   (select_expression
    (term
     value: 
      (invocation
       (object_reference name: (identifier))
       (
       parameter: (term value: (literal))
       parameter: ,
       (term
        value: 
         (invocation
          (object_reference name: (identifier))
          (
          parameter: (term value: (literal))
          parameter: ,
          (term value: (literal))
          )))
       ))))))
 ;)
```

Reply:
Is it what you meant:

```sql
CREATE FUNCTION public.update_time() RETURNS trigger
  LANGUAGE plpgsql
AS $$
BEGIN
  RETURN $$abc$$;
END;
$$;
```
It parse fine:
```
(program
 (statement
  alias: 
   (create_function (keyword_create) (keyword_function)
    (object_reference schema: (identifier) . name: (identifier))
    ( ) (keyword_returns) (keyword_trigger)
    (function_language (keyword_language) (identifier))
    (function_body (keyword_as) (dollar_quote) (keyword_begin) (keyword_return)
                   (literal)
     ; (keyword_end) ; (dollar_quote))))
 ;)
```

Reply:
> Is it what you meant:
>
> ```sql
> CREATE FUNCTION public.update_time() RETURNS trigger
>   LANGUAGE plpgsql
> AS $$
> BEGIN
>   RETURN $$abc$$;
> END;
> $$;
> ```

right, ideally it shouldn't because the real parser sees the second $$ right after `RETURN`, thinks the function body is complete, and doesn't know what to do with the rest of the input:

```
ERROR:  syntax error at or near "abc$$"
LINE 5:   RETURN $$abc$$;
```

your first example is fine because each of the strings is separate, there's no nesting. I'm thinking of something more like your second, or like this case:

```sql
create table if not exists mytable (id serial primary key, val text);
do $a$
begin
  execute format(
    $b$
    -- note the escaped %% in the "c" format!
    insert into %1$s (val)
    select format($c$%%1$L$c$, 'alpha');
    $b$,
    'mytable'
  );
end
$a$;
select * from mytable;
```

I think your implementation is a big improvement on what we've got! but it's more permissive than Postgres' actual parser in a way that is likely to be a trap for the unwary (and it's split between this dollar_quote and the external dollar_string, so even more complicated) so I think we've got to make that really clear. I'll add a more detailed explanation to scanner.c.

```suggestion
    // dollar_quote is only used in the `create function` statement. Postgres-style
    // dollar quoting in other contexts is done by the external _dollar_string. See
    // scanner.c, and especially the note about parsing of nested dollar quotes.
```

Reply:
My scanner was a simpler version of https://github.com/m-novikov/tree-sitter-sql/blob/main/src/scanner.cc , which produce 3 tokens , start tag, end tag and string content.
Maybe it is a better choice for nested dollar strings with the same name inside a body function.
And also for language injection https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection.